src/langchain_google_firestore/document_loader.py

```diff
@@ -0,0 +1,247 @@
+    ) -> None:
+        """Document Saver for Google Cloud Firestore.
+        Args:
+            collection: The single `/`-delimited path to a Firestore collection. If this
+              value is present it will write documents with an auto generated id.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        try:
+            from google.cloud import firestore
+            from google.cloud.firestore_v1.services.firestore.transports.base import (
+                DEFAULT_CLIENT_INFO,
+            )
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        self.collection = collection
+        client_info = DEFAULT_CLIENT_INFO
+        client_info.user_agent = USER_AGENT
+        if client:
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            self.client = firestore.Client(client_info=client_info)
+
+    def upsert_documents(
+        self,
+        documents: List[Document],
+        merge: Optional[bool] = False,
+        document_ids: Optional[List[str]] = None,
+    ) -> None:
+        """Create / merge documents into the Firestore database.
+        Args:
+         documents: List of documents to be written into Firestore.
+         merge: To merge data iwth an existing document (creating if the document does
+          not exist).
+         document_ids: List of document ids to be used. By default it will try to
+          construct the document paths using the `reference` from the Document.
+        """
+        try:
+            from google.cloud.firestore_v1.document import DocumentReference
+        except ImportError:
+            raise ImportError(IMPORT_ERROR_MSG)
+
+        db_batch = self.client.batch()
+
+        if document_ids and (len(document_ids) != len(documents)):
+            raise ValueError("Document ids and docs must have the same size")
+
+        if document_ids:
+            docs_list = tuple(zip(documents, document_ids))
+        else:
+            docs_list = documents
+
+        for batch in self._batched(docs_list, WRITE_BATCH_SIZE):
+            for elem in batch:
+                if document_ids:
+                    doc = elem[0]
+                    doc_id = elem[1]
+                else:
+                    doc = elem
+                    doc_id = None
+                document_dict = DocumentConverter.convertLangChainDocument(
+                    doc, self.client
+                )
+                if self.collection:
+                    doc_ref = self.client.collection(self.collection).document()
+                elif doc_id:
+                    doc_ref = DocumentReference(*doc_id.split("/"), client=self.client)
+                elif document_dict["path"]:
+                    doc_ref = DocumentReference(
+                        *document_dict["path"].split("/"), client=self.client
+                    )
+                else:
+                    continue
```

Comment:
In this case we were unable to get a valid Document Reference.

Either because, there's no collection set by Cx, nor valid doc_id (they could be sending empty strings or None), nor path in the document_dict.

We can't insert that document. We either fail all the insert (which could be thousands of documents) or skip it.
IIRC I mentioned I would skip it if this happened in the design doc.

I can raise an error if you prefer, I'm not certain what is better here. Some documents could be already committed in either case. And either the Cx tries to figure out what documents where inserted, then remove them and retry.